### PR TITLE
feat: Added github actions for keg-hub

### PR DESCRIPTION
**Pull request by lancetipton**
**Ticket**: [ZEN-392](https://jira.simpleviewtools.com/browse/ZEN-392)

[Semver-Status] minor

## Context

* We need to setup CI with github actions to automate common tasks

## Goal

* Setup some inital github actions for common tasks
* The following actions were added
  * **Build** - Runs on `push` to `master`
  * **Format** - Runs on `push` to `develop`
  * **PR-Changes** - Runs on **any** `pull_request`
  * **Tests** - Runs on **any** `pull_request`
* Added helper scripts for running the actions

## Updates
* `.github/workflows/*`
  * Github workflow files to run actions
* `scripts/ci/*`
  * Helper scripts run by the github actions

## Testing

**Setup**
* Fork the `keg-hub` repo into you're own account
* Pull the fork locally, or update your local repo to point to your fork
* **All the following tests should use your forked repo!**

**Test 1**
* After ensuring your git origin is poiting to your fork, switch to the `keg-github-actions` branch
* Next modify some files in different ways in a few different repos
  * For example, add a file to `Re-Theme`, and delete a file in `Keg-Core`
  * **IMPORTANT** - Do not modify all repos. We want to test that only the changed repos are effected
* Commit the changes you made, and push them to github.
  * Again make sure you push to **your own forked repo**
* After pushing to github, ensure no actions run

**Test 2**
* Create a pull request from the recently pushed branch ( that has your changes ) into **your forked version's** develop branch
* At this point, two of the github actions should run
  * `keg-test` and `keg-pr-changes`
  * Ensure both actions run and succeed
  * The `keg-test` workflow might fail depending on your changes
    * If it does, you've validated that it works properly ( good job )
    * Go revert the changes you made to brake the tests, and start over
    * 
  * The `keg-pr-changes`
    * This workflow shows a warning about invalid `inputs`
    * This is an issue with the action.yml file, that the creator needs to update
    * It's just a warning though, and should work as expected
* Assuming both workflows have succeeded
  * Go to the github action called `Run Repo Tests`
    * Toggle open the `Run Repo Tests` step
      * Ensure tests were only run for the repos that were changed
        * [Looks like this](http://snpy.in/5TUg5E)
  * Next, go to the PR's discussion tab
    * You should see a comment with a list of all changed files separated by repo
      * [Looks like this](http://snpy.in/67bObv)
    * Ensure only the repos that had changes are listed in the auto-generated comment

**Test 3**
* Merge the changes from you branch into the develop branch of **your fork**
  * **IMPORTANT** - Be sure it's your fork, and not the simpleviewinc one
* This should kick off the `keg-format` workflow
  * Go to the actions tab, and validate that the workflow ran successfully
  * [Looks like this](http://snpy.in/b5ORTB)
* Next goto the commit history for the develop branch
* Ensure a new commit was added with the formatted code
  * [Looks like this](http://snpy.in/Y9tMrL)

**Test 4**
* Finally, merge **your** develop branch into **your** the master branch
  * You'll probably have to do it from the command line
  * Make sure you pull down the changes to develop first
    * These changes are from the `keg-format` workflow of **Test 3**
* This should kick off the `keg-build` workflow
  * It will run tests, then build each one of the repos that has changed files
  * [Looks like this](http://snpy.in/a9u91S)
* It will then commit those build changes back into the master branch
  * Ensure a new commit was added with the built code
  * [Looks like this](http://snpy.in/VyudLV)

### CODE REVIEW

File: .github/workflows/keg-build.yml
Comment (daniel112) [inline .github/workflows/keg-build.yml:41]: when this workflow executes. `master` will be 1 ahead from `develop` which might cause conflict on the next push from develop > master

just thinking out loud here; maybe we can:
- add another step to merge down master to develop after the commit?
- update the build step to be on develop instead of master. master can be just for publishing on push?

**Review by mpcarolin — APPROVED** [review]
Looks good. This will be so helpful 👍🏼

File: scripts/ci/getChangedFiles.sh
Comment (daniel112) [inline scripts/ci/getChangedFiles.sh:12]: on the `build` workflow, im running into an issue where it's not actually building any of the repos.

this commit should have some build file updates for jsutils https://github.com/daniel112/keg-hub/commit/083e5bad43abbc7bbc214e5cca622b91a3799245

here are the file changes when I run `yarn build` on jsutils locally
- [image](https://user-images.githubusercontent.com/3317835/103705518-671bcb80-4f68-11eb-8d76-3db74e1f828f.png)

if I understand this right, when it's on the `master` branch, it's comparing itself to `develop`, so it wont actually find any file changes

**Review by daniel112 — COMMENTED** [review]
all the tests work as defined in the test instructions. but I have some comments and questions on some of them

File: .github/workflows/keg-format.yml
Comment (lancetipton) [reply .github/workflows/keg-format.yml:33]: * It **should** not be possible for the linter to break our code. 
* Keyword is should, cause anything is possible, the whole point of a linter is to enforce a correct coding standard.
  * I've never heard of a linter that formats code in a way that breaks it. What would be the point?
* So yes, it's possible but pretty sure most the internet would break as well.
  * Prettier / ESLint are _almost_ a requirement these days for most production grade repositories.

Honestly, I think it's a non-issue

File: .github/workflows/keg-build.yml
Comment (lancetipton) [reply .github/workflows/keg-build.yml:41]: I like it, let me look into it

File: .github/workflows/keg-format.yml
Comment (daniel112) [reply .github/workflows/keg-format.yml:33]: i figured it would be a bit of a stretch. works for me

File: .github/workflows/keg-build.yml
Comment (lancetipton) [reply .github/workflows/keg-build.yml:41]: Ok, update keg-build action to run on pushes to develop

File: scripts/ci/getChangedFiles.sh
Comment (lancetipton) [reply scripts/ci/getChangedFiles.sh:12]: This should be resolved now that the build action runs on pushes to develop

File: .github/workflows/keg-build.yml
Comment (daniel112) [inline .github/workflows/keg-build.yml:29]: I will re-test again tomorrow. but after testing it tonight here's what I found: 
1. if the `format/test/build` steps fail. it will still count it as **success**
    - see [this](https://github.com/daniel112/keg-hub/actions/runs/464965447) workflow for when `format` step throws an error
2. jsutils  fails at yarn build  when on node 14.15.1  (default)
      - see `run repo build` step: https://github.com/daniel112/keg-hub/runs/1653846780?check_suite_focus=true
      - node v 14.15.1 is default for `ubuntu-latest` runner
      - this is the error when running `yarn build` on  node `14.15.1` - [image](https://user-images.githubusercontent.com/3317835/103721867-fbe3f080-4f8b-11eb-8cbd-90a3f105de10.png)
      - I manage to bypass this by setting the node version to `12` on the workflow (since we currently use this major version)
      - ```js
                   - name: "Setup node Version"
                      uses: actions/setup-node@v2
                      with:
                        node-version: '12.x.x'
         ```
      - see the workflow that successfully builds jsutils [here](https://github.com/daniel112/keg-hub/runs/1653917635?check_suite_focus=true)

Comment (lancetipton) [reply .github/workflows/keg-build.yml:34]: * Updated bash scripts to exit on error
* Updated all workflows to run on version 12 of node
 * Updated Keg-Components and Keg-Core to pass formatter script
 * ReTheme needs a number of updates so holding off on that one

**Review by daniel112 — APPROVED** [review]
will approve since everything seems to be working great
but as discussed,
 we need to wait until re-theme lint errors are fixed and merged in before merging this one

**Pull request merged**
